src/include/sof/audio/component.h

```diff
@@ -335,7 +335,8 @@ struct comp_dev *comp_new(struct sof_ipc_comp *comp);
  */
 static inline void comp_free(struct comp_dev *dev)
 {
-	dev->drv->ops.free(dev);
+	if (dev->drv->ops.free)
```

Comment:
_comp ops_, as an application level extension point, might be the first one and most popular area of the development based on SOF. That is why there should be strong verification of the API requirements enforced I believe. The documentation above says that `ops.free()` is mandatory which seems right since it might be better to enforce empty handler rather than allow for memory leaks. The proposed change quietly allows to skip the implementation which is inconsistent with the documented requirements. This one (as well as other handlers marked as mandatory like `params()`) seems to be a good candidate for `assert(not null)`. And I would opt for assert() that compiles in non-debug version to panic (it might be difficult to deliver extra debug configuration on some platforms and force the use) to make sure that no mandatory part is missed during the component development process.

Reply:
@mmaka1  Ack, if it's mandatory we shout and let the user know. If optional we perform the check.